crates/re_data_source/src/load_file.rs

```diff
@@ -1,134 +1,255 @@
+
+    let is_builtin = is_builtin(path, is_dir);
+    let is_rrd = crate::SUPPORTED_RERUN_EXTENSIONS.contains(&extension(path).as_str());
+
+    (!is_rrd && is_builtin).then(|| {
+        LogMsg::SetStoreInfo(SetStoreInfo {
+            row_id: re_log_types::RowId::new(),
+            info: re_log_types::StoreInfo {
+                application_id: app_id.clone(),
+                store_id: store_id.clone(),
+                is_official_example: false,
+                started: re_log_types::Time::now(),
+                store_source,
+                store_kind: re_log_types::StoreKind::Recording,
+            },
+        })
+    })
+}
+
+/// Loads the data at `path` using all available [`crate::DataLoader`]s.
+///
+/// Returns a channel with all the [`LoadedData`]:
+/// - On native, this is filled asynchronously from other threads.
+/// - On wasm, this is pre-filled synchronously.
+#[cfg_attr(target_arch = "wasm32", allow(clippy::needless_pass_by_value))]
+fn load(
+    store_id: &re_log_types::StoreId,
+    path: &std::path::Path,
+    is_dir: bool,
+    contents: Option<std::borrow::Cow<'_, [u8]>>,
+) -> Result<std::sync::mpsc::Receiver<LoadedData>, DataLoaderError> {
+    let extension = extension(path);
+    let is_builtin = is_builtin(path, is_dir);
+
+    if !is_builtin {
+        return if extension.is_empty() {
+            Err(anyhow::anyhow!("files without extensions (file.XXX) are not supported").into())
+        } else {
+            Err(anyhow::anyhow!(".{extension} files are not supported").into())
+        };
+    }
+
+    // On native we run loaders in parallel so this needs to become static.
+    #[cfg(not(target_arch = "wasm32"))]
+    let contents: Option<Arc<std::borrow::Cow<'static, [u8]>>> =
+        contents.map(|contents| Arc::new(Cow::Owned(contents.into_owned())));
+
+    let rx_loader = {
+        let (tx_loader, rx_loader) = std::sync::mpsc::channel();
+
+        for loader in crate::iter_loaders() {
+            let mut contents = contents.clone(); // arc
+            let loader = Arc::clone(&loader);
+            let store_id = store_id.clone();
+            let tx_loader = tx_loader.clone();
+            let path = path.to_owned();
+
+            spawn(move || {
+                if let Some(contents) = contents.take() {
+                    if let Err(err) = loader.load_from_file_contents(
+                        store_id,
+                        path.clone(),
+                        Cow::Borrowed(&contents),
+                        tx_loader,
+                    ) {
+                        re_log::error!(?path, loader = loader.name(), %err, "failed to load data from file");
+                    }
+                } else {
+                    #[cfg(not(target_arch = "wasm32"))]
+                    if let Err(err) = loader.load_from_file(store_id, path.clone(), tx_loader) {
+                        re_log::error!(?path, loader = loader.name(), %err, "failed to load data from files");
```

Comment:
Small nit, but you should capitalize the first letter in log messages since they are actually part of the ui (`re_log::error`s are shown as popups to the user)